Homework 04

art-frela opened the pull request:
Task 1 - Described **Storager{}** interface and implemented that, as JSON and XML storage.
Task 2 - work with JSON storage, implement Sort{} interface
Task 3 - Added calculator package method HelpPrint () and converted to sin & cos with degree arguments, removed other trigonometric functions
Task 4 - For that disposition: ![chees_field](https://user-images.githubusercontent.com/32013768/57135206-41ace680-6db1-11e9-9dc8-97aab7eb0d3e.png)
print possible moves for kNights (horses)

art-frela committed:
homework-04

implement mandatory part
tasks 1-3

art-frela committed:
fix with errors

want to sleep )

art-frela committed:
task#4 implement

simple version witout mirroring blac/white
has too much double of code

art-frela committed:
add some comments

describe packages

tirava commented on homework-04/homework-04.go:
```diff
@@ -0,0 +1,315 @@
+	"encoding/xml"
+	"fmt"
+	"gocourse221/calculator"
+	"gocourse221/chees"
```
Скорее всего у Сергея код не запустится без плясок, да и мне пришлось поковыряться, т.к. структура проекта явно заточена под твои локальные папки
Здесь лучше использовать "_./chees_", а сборки хранить в подпапках относительно _main.go_, а не _src_

art-frela replied:
дык делал и так и этак... не хочет никак зараза.

tirava commented on homework-04/homework-04.go:
```diff
@@ -0,0 +1,315 @@
+		filename: userxml,
+	}
+	jStorage.readStorage()
+	xStorage.readStorage()
```
Нет обработки ошибок для стораджей - опасненько

tirava commented on homework-04/homework-04.go:
```diff
@@ -0,0 +1,315 @@
+	filecontent, _ := ioutil.ReadAll(f)
+	err = json.Unmarshal(filecontent, &storage.data)
+	return
+}
```
Странная реализация чтения, либо я что-то не понял, и defer лишний, т.к.  ReadAll атомарная функция (сама все делает от открытия до закрытия)
Я бы сделал так:
```
r := strings.NewReader
b, err := ioutil.ReadAll(r)
```

art-frela replied:
Посмотрел в исходние [ioutil](https://golang.org/src/io/ioutil/ioutil.go) там нет закрытия файла. Так что defer тут по делу, но реализация не типичная согласен (когда-то давно сделал сдуру, работает, вот и лепдю везде)

tirava commented on homework-04/homework-04.go:
```diff
@@ -0,0 +1,315 @@
+}
+
+// readStorage - read data from file
+func (storage storageJSONfile) saveStorage() (err error) {
```
А почему здесь не по ссылке _storageJSONfile_, а выше чтение было по ссылке

art-frela replied:
потому что чтение вносит изменение в сам объект, а запись может выполниться и по копии значения

tirava replied:
Вообще, странно звучит:

> чтение вносит изменение в сам объект, а запись может выполниться и по копии значения

Я бы сказал, что в реальности все наоборот, нет ли тут косяка в логике..

art-frela replied:
да да.. наоброт конечно же )) 
что-то переклинило меня при написании данного ответа )

tirava commented on homework-04/homework-04.go:
```diff
@@ -0,0 +1,315 @@
+}
+
+// PhoneBook - type of slice Users
+type PhoneBook []User
```
Долго я искал эту строку )
Я бы перенес куда-то под константы

tirava commented on homework-04/homework-04.go:
```diff
@@ -0,0 +1,315 @@
+Date: 2019-05-03
+Task:
+1. Написать свой интерфейс и создать несколько структур, удовлетворяющих ему.
+2. Создать псевдоним типа телефонной книги и реализовать для него интерфейс Sort{}.
```
Молодец, в целом понравилась реализация первых двух заданий.

tirava commented on homework-04/homework-04.go:
```diff
@@ -0,0 +1,315 @@
+Task:
+1. Написать свой интерфейс и создать несколько структур, удовлетворяющих ему.
+2. Создать псевдоним типа телефонной книги и реализовать для него интерфейс Sort{}.
+3. Дописать функцию, которая будет выводить справку к калькулятору. Она должна вызываться при вводе слова help вместо выражения.
```
Калькулятор не смотрел, и сам его не делал, т.к. это задание странное:
1. Расширение какого-то калькулятора, которого мы и не видели раньше
2. Фактически задание - копипаст кода из методички с добавлением небольшой функции

tirava commented on homework-04/homework-04.go:
```diff
@@ -0,0 +1,315 @@
+3. Дописать функцию, которая будет выводить справку к калькулятору. Она должна вызываться при вводе слова help вместо выражения.
+4. * Написать функцию, которая будет получать позицию коня на шахматной доске, а возвращать массив из точек, в которые конь сможет сделать ход.
+	Точку следует обозначить как структуру, содержащую x и y типа int
+	Получение значений и их запись в точку должны происходить только с помощью отдельных методов. В них надо проводить проверку на то, что такая точка может существовать на шахматной доске.
```
К сожалению, у меня нет времени сделать 4-е задание, там явно не 1 час нужен, хотя хотелось бы
Просто запустил, глянуть как у тебя...
Молодец, видно, что большая работа сделана (интересно, за сколько времени), но...
...судя по заданию, нужно было всего лишь получить позицию коня и выдать возможные ходы, а про другие фигуры на доске не сказано,  и не сказано, что коняшки стоят в фикс. позициях сразу после старта
Плюс названия ходов какие-то не шахматные что ли, в общем мне не хватило более простой реализации, например:
```
Укажите позицию коня: d4
Возможные ходы: {f5, f3, c6, ...}
```
А то что там сказано про обозначить как структуру _int x,y_ - ну так это под капотом и будет

art-frela replied:
да я начал было более общую задачу решать... для всей типа игры, потом как АПИ можно сделать к вэбу, но всплыло направление ходов, съедания фигур соперника... в общем решил остановиться. А по координатам позиций это да... я еле себя заставил переделать в цифры как требуется в тз. 

> "Точку следует обозначить как структуру, содержащую x и y типа int" 
вот и сделал так странно... вообще задание оторвано от реальности немного, наверное в попытке упростить.

art-frela committed:
implement task#1

use functions with using interface
and replace packages to local folder
fix methods - all work with pointers

arcsub commented on homework-04/chees/chees.go:
```diff
@@ -0,0 +1,255 @@
+	armies := [2]cheesArmy{whiteArmy, blackArmy}
+
+	chees.Armies = armies
+}
```
Отлично! Действительно, очень большую работу проделал, это видно ! Рад что у тебя нашлось на время на это задание! Код хорош, видно что знания растут 👍
Замечаний по коду особо нет, разве что вместо стринговых значений цвета я бы всё таки использовал константы

arcsub commented on homework-04/homework-04.go:
```diff
@@ -0,0 +1,375 @@
+	}
+	err := jStorage.readStorage()
+	errHandler("read JSON storage", err)
+
```
- Названия переменных в некоторых местах не очень удачны, всё же лучше бы они смотрелись без сокращений типа `xStorage`, а полноценно `XMLStorage`, но это чисто моё субъективное мнение

arcsub commented on homework-04/homework-04.go:
```diff
@@ -0,0 +1,375 @@
+			uid = usr.UID
+		}
+	}
+	uid++ //next ID
```
- Для нашего начального пути в Go-Developer'ы такой код конечно же подойдёт, но тут всё же лучше использовать карту типа `map[uuid]User`, вместо среза `[]User`, и не пробегаться по срезу каждый раз в поисках нового ID - лучше юзать генератор UUID, делается легко.
Аналогичная ситуация и с функцией `GetByID()` - мы могли бы её сильно упростить, не прибегая к конструкции `for`:
```
func (storage *storageJSONfile) GetByID(uuid UUID) (user User, err error) {
    if user, ok := storage.Users[uuid]; !ok {
        err = fmt.Errorf("User with ID=%d is't exist", uid)
    }
    return
```
что то вроде этого
